gcloud/monitoring/resource.py

```diff
@@ -0,0 +1,160 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Resource Descriptors for the Google Monitoring API."""
+
+import collections
+
+from gcloud.monitoring.label import LabelDescriptor
+
+
+class ResourceDescriptor(collections.namedtuple(
+        'ResourceDescriptor', ('name type display_name description labels'))):
+    """Specification of a monitored resource type and its schema.
+
+    Resource descriptor instances are immutable.
+
+    :type name: string
+    :param name:
+        The "resource name" of the monitored resource descriptor:
+        ``"projects/<project_id>/monitoredResourceDescriptors/<type>"``
+
+    :type type: string
+    :param type:
+        The monitored resource type. For example: ``"gce_instance"``
+
+    :type display_name: string
+    :param display_name:
+        A concise name that might be displayed in user interfaces.
+
+    :type description: string
+    :param description:
+        A detailed description that might be used in documentation.
+
+    :type labels: list of :class:`~gcloud.monitoring.label.LabelDescriptor`
+    :param labels:
+        A sequence of label descriptors specifying the labels used
+        to identify a specific instance of this monitored resource.
+    """
+    __slots__ = ()
+
+    @classmethod
+    def _fetch(cls, client, resource_type):
+        """Look up a resource descriptor by type.
+
+        :type client: :class:`gcloud.monitoring.client.Client`
+        :param client: The client to use.
+
+        :type resource_type: string
+        :param resource_type: The resource type name.
+
+        :rtype: :class:`ResourceDescriptor`
+        :returns: The resource descriptor instance.
+
+        :raises: :class:`gcloud.exceptions.NotFound`
+        """
+        path = '/projects/{}/monitoredResourceDescriptors/{}'.format(
+            client.project, resource_type)
+        info = client.connection.api_request('GET', path)
+        return cls._from_dict(info)
+
+    @classmethod
+    def _list(cls, client, filter=None):
+        """List all resource descriptors for the project.
+
+        :type client: :class:`gcloud.monitoring.client.Client`
+        :param client: The client to use.
+
+        :type filter: string or None
+        :param filter: An optional filter string describing the resource
+                       descriptors to be returned.
+
+        :rtype: list of :class:`ResourceDescriptor`
+        :returns: A list of resource descriptor instances.
+        """
+        # Allow "filter" as a parameter name: pylint: disable=redefined-builtin
+
+        path = '/projects/{}/monitoredResourceDescriptors/'.format(
+            client.project)
+
+        def _descriptors():
```

Comment:
It only exists to be called once. This is just a way of structuring the logic for constructing the result list. It would not be appropriate to give the generator a signature that pretends to be meaningful in a broader context.